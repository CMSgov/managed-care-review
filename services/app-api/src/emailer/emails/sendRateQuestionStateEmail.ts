import type { RateQuestionType, RateType } from '../../domain-models'
import type { EmailConfiguration, EmailData } from '../emailer'
import {
    getActuaryContactEmails,
    getRateStateContactEmails,
    getRateSubmitterEmails,
    renderTemplate,
    stripHTMLFromTemplate,
} from '../templateHelpers'
import { formatCalendarDate } from '../../common-code/dateHelpers'
import { pruneDuplicateEmails } from '../formatters'
import { rateQuestionResponseURL } from '../generateURLs'

export const sendRateQuestionStateEmail = async (
    rate: RateType,
    config: EmailConfiguration,
    rateQuestion: RateQuestionType
): Promise<EmailData | Error> => {
    const rateFormData = rate.packageSubmissions[0].rateRevision.formData
    const parentContractID = rate.parentContractID
    const shouldIncludeActuaries =
        rateFormData.actuaryCommunicationPreference === 'OACT_TO_ACTUARY'

    if (!rateFormData.rateCertificationName) {
        return new Error(
            'Error getting rate name. Rate certification name was undefined.'
        )
    }

    const submitterEmails = getRateSubmitterEmails(rate)
    const stateContactEmails = getRateStateContactEmails(rate)
    const actuaryEmails = shouldIncludeActuaries
        ? getActuaryContactEmails(rate)
        : []

    const toAddresses = pruneDuplicateEmails([
        ...submitterEmails,
        ...stateContactEmails,
        ...actuaryEmails,
    ])

    const data = {
        packageName: rateFormData.rateCertificationName,
        questionResponseURL: rateQuestionResponseURL(
            parentContractID,
            rateQuestion.rateID,
            config.baseUrl
        ),
        cmsRequestorEmail: rateQuestion.addedBy.email,
        cmsRequestorName: `${rateQuestion.addedBy.givenName} ${rateQuestion.addedBy.familyName}`,
        cmsRequestorDivision: rateQuestion.division,
        dateAsked: formatCalendarDate(
            rateQuestion.createdAt,
            'America/Los_Angeles'
        ),
    }

    const template = await renderTemplate<typeof data>(
        'sendQuestionStateEmail',
        data
    )

    if (template instanceof Error) {
        return template
    } else {
        return {
            toAddresses: toAddresses,
            ccAddresses: [...config.dmcpSubmissionEmails],
            replyToAddresses: [],
            sourceEmail: config.emailSource,
            subject: `${
                config.stage !== 'prod' ? `[${config.stage}] ` : ''
            }New questions about ${rateFormData.rateCertificationName}`,
            bodyText: stripHTMLFromTemplate(template),
            bodyHTML: template,
        }
    }
}
