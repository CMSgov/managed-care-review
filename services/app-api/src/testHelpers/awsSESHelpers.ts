import { EmailData } from '../emailer'
import { SESServiceException } from '@aws-sdk/client-ses'

async function testSendSESEmail(
    emailData: EmailData
): Promise<string | SESServiceException> {
    console.info('SendEmailCommand: ')
    return 'Email content' + JSON.stringify(emailData)
}

export { testSendSESEmail }
