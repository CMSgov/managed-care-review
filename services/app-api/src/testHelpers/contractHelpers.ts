import { InsertContractArgsType } from '../postgres/contractAndRates/insertContract'

const createDraftContractData = (
    contractArgs?: Partial<InsertContractArgsType>
): InsertContractArgsType => {
    return {
        stateCode: contractArgs?.stateCode ?? 'MN',
        submissionType: contractArgs?.submissionType ?? 'CONTRACT_AND_RATES',
        submissionDescription:
            contractArgs?.submissionDescription ?? 'Contract 1.0',
        contractType: contractArgs?.contractType ?? 'BASE',
        programIDs: contractArgs?.programIDs ?? ['PMAP'],
        populationCovered: contractArgs?.populationCovered ?? 'MEDICAID',
        riskBasedContract: contractArgs?.riskBasedContract ?? false,
    }
}

export { createDraftContractData }
