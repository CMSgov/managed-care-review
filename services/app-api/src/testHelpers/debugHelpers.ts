import util from 'util'

const consoleLogFullData = (data: unknown) =>
    console.info(util.inspect(data, false, null, true))

export { consoleLogFullData }
