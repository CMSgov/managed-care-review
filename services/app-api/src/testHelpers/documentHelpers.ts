import type { GenericDocument } from '../gen/gqlServer'

// Clear document metadata that is added by API
// enables direct compare between document lists (created by FE and submitted by users)
// and the return  document lists (created by BE parsers and API and include metadata such as dateAdded)
const clearDocMetadata = (documents?: GenericDocument[]): GenericDocument[] => {
    if (!documents) return []
    return documents.map((doc) => {
        // clear out metadata fields
        delete doc['dateAdded']
        return doc
    })
}

export { clearDocMetadata }
