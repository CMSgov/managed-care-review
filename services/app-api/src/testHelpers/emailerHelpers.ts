import {
    EmailConfiguration,
    EmailData,
    Emailer,
    newPackageCMSEmail,
    newPackageStateEmail,
    UpdatedEmailData,
    unlockPackageCMSEmail,
    unlockPackageStateEmail,
    resubmittedStateEmail,
    resubmittedCMSEmail,
} from '../emailer'
import {
    LockedHealthPlanFormDataType,
    UnlockedHealthPlanFormDataType,
} from '../../../app-web/src/common-code/healthPlanFormDataType'
import { UserType, StateUserType } from '../domain-models'

const testEmailConfig: EmailConfiguration = {
    stage: 'LOCAL',
    baseUrl: 'http://localhost',
    emailSource: 'emailSource@example.com',
    cmsReviewSharedEmails: ['cmsreview1@example.com', 'cmsreview2@example.com'],
    cmsMcogEmailAddress: 'mcog@example.com',
    cmsRateEmailAddress: 'rates@example.com',
    cmsDirectReviewTeamEmailAddress: 'mc-review@example.com',
    ratesReviewSharedEmails: ['ratesreview@example.com'],
}

const submissionName = 'MN-PMAP-0001'

const testEmailer = (customConfig?: EmailConfiguration): Emailer => {
    const config = customConfig || testEmailConfig
    return {
        sendEmail: jest.fn(
            async (emailData: EmailData): Promise<void | Error> => {
                console.log('Email content' + JSON.stringify(emailData))
            }
        ),
        sendCMSNewPackage: function async(
            submission: LockedHealthPlanFormDataType,
            submissionName: string
        ): Promise<void | Error> {
            const emailData = newPackageCMSEmail(
                submission,
                submissionName,
                config
            )
            return this.sendEmail(emailData)
        },
        sendStateNewPackage: function async(
            submission: LockedHealthPlanFormDataType,
            submissionName: string,
            user: UserType
        ): Promise<void | Error> {
            const emailData = newPackageStateEmail(
                submission,
                submissionName,
                user,
                config
            )
            return this.sendEmail(emailData)
        },
        sendUnlockPackageCMSEmail: function async(
            submission: UnlockedHealthPlanFormDataType,
            updatedEmailData: UpdatedEmailData,
            submissionName: string
        ): Promise<void | Error> {
            const emailData = unlockPackageCMSEmail(
                submission,
                updatedEmailData,
                config,
                submissionName
            )
            return this.sendEmail(emailData)
        },
        sendUnlockPackageStateEmail: function async(
            submission: UnlockedHealthPlanFormDataType,
            updatedEmailData: UpdatedEmailData
        ): Promise<void | Error> {
            const emailData = unlockPackageStateEmail(
                submission,
                updatedEmailData,
                config,
                submissionName
            )
            return this.sendEmail(emailData)
        },
        sendResubmittedStateEmail: function async(
            submission: LockedHealthPlanFormDataType,
            updatedEmailData: UpdatedEmailData,
            user: UserType
        ): Promise<void | Error> {
            const emailData = resubmittedStateEmail(
                submission,
                user,
                updatedEmailData,
                config
            )
            return this.sendEmail(emailData)
        },
        sendResubmittedCMSEmail: function async(
            submission: LockedHealthPlanFormDataType,
            updatedEmailData: UpdatedEmailData
        ): Promise<void | Error> {
            const emailData = resubmittedCMSEmail(
                submission,
                updatedEmailData,
                config
            )
            return this.sendEmail(emailData)
        },
    }
}

const mockUser = (): StateUserType => {
    return {
        role: 'STATE_USER',
        email: 'test+state+user@example.com',
        name: 'Test State User',
        state_code: 'MN',
    }
}

const mockContractAndRatesFormData = (
    submissionPartial?: Partial<LockedHealthPlanFormDataType>
): LockedHealthPlanFormDataType => {
    return {
        createdAt: new Date(),
        updatedAt: new Date(),
        status: 'SUBMITTED',
        stateNumber: 3,
        id: 'test-abc-125',
        stateCode: 'MN',
        programIDs: ['snbc'],
        submissionType: 'CONTRACT_AND_RATES',
        submissionDescription: 'A submitted submission',
        submittedAt: new Date(),
        documents: [
            {
                s3URL: 'bar',
                name: 'foo',
                documentCategories: ['RATES_RELATED' as const],
            },
        ],
        contractType: 'BASE',
        contractExecutionStatus: 'EXECUTED',
        contractDocuments: [
            {
                s3URL: 'bar',
                name: 'foo',
                documentCategories: ['CONTRACT' as const],
            },
        ],
        contractDateStart: new Date(),
        contractDateEnd: new Date(),
        managedCareEntities: ['ENROLLMENT_PROCESS'],
        federalAuthorities: ['VOLUNTARY', 'BENCHMARK'],
        rateType: 'NEW',
        rateDocuments: [
            {
                s3URL: 'bar',
                name: 'foo',
                documentCategories: ['RATES' as const],
            },
        ],
        rateDateStart: new Date(),
        rateDateEnd: new Date(),
        rateDateCertified: new Date(),
        rateAmendmentInfo: null,
        stateContacts: [
            {
                name: 'Test Person',
                titleRole: 'A Role',
                email: 'test+state+contact@example.com',
            },
        ],
        actuaryContacts: [],
        ...submissionPartial,
    }
}

const mockUnlockedContractAndRatesFormData = (
    submissionPartial?: Partial<UnlockedHealthPlanFormDataType>
): UnlockedHealthPlanFormDataType => {
    return {
        createdAt: new Date(),
        updatedAt: new Date(),
        status: 'DRAFT',
        stateNumber: 3,
        id: 'test-abc-125',
        stateCode: 'MN',
        programIDs: ['snbc'],
        submissionType: 'CONTRACT_AND_RATES',
        submissionDescription: 'A submitted submission',
        documents: [
            {
                s3URL: 'bar',
                name: 'foo',
                documentCategories: ['RATES_RELATED' as const],
            },
        ],
        contractType: 'BASE',
        contractExecutionStatus: 'EXECUTED',
        contractDocuments: [
            {
                s3URL: 'bar',
                name: 'foo',
                documentCategories: ['CONTRACT' as const],
            },
        ],
        contractDateStart: new Date(),
        contractDateEnd: new Date(),
        managedCareEntities: ['ENROLLMENT_PROCESS'],
        federalAuthorities: ['VOLUNTARY', 'BENCHMARK'],
        rateType: 'NEW',
        rateDocuments: [
            {
                s3URL: 'bar',
                name: 'foo',
                documentCategories: ['RATES' as const],
            },
        ],
        rateDateStart: new Date(),
        rateDateEnd: new Date(),
        rateDateCertified: new Date(),
        rateAmendmentInfo: undefined,
        stateContacts: [
            {
                name: 'Test Person',
                titleRole: 'A Role',
                email: 'test+state+contact@example.com',
            },
        ],
        actuaryContacts: [],
        ...submissionPartial,
    }
}

const mockUnlockedContractOnlyFormData = (
    submissionPartial?: Partial<UnlockedHealthPlanFormDataType>
): UnlockedHealthPlanFormDataType => {
    return {
        createdAt: new Date(),
        updatedAt: new Date(),
        status: 'DRAFT',
        stateNumber: 3,
        id: 'test-abc-125',
        stateCode: 'MN',
        programIDs: ['snbc'],
        submissionType: 'CONTRACT_ONLY',
        submissionDescription: 'A submitted submission',
        documents: [
            {
                s3URL: 'bar',
                name: 'foo',
                documentCategories: ['CONTRACT_RELATED' as const],
            },
        ],
        contractType: 'BASE',
        contractExecutionStatus: 'EXECUTED',
        contractDocuments: [
            {
                s3URL: 'bar',
                name: 'foo',
                documentCategories: ['CONTRACT' as const],
            },
        ],
        contractDateStart: new Date(),
        contractDateEnd: new Date(),
        managedCareEntities: ['ENROLLMENT_PROCESS'],
        federalAuthorities: ['VOLUNTARY', 'BENCHMARK'],
        rateDocuments: [],
        stateContacts: [
            {
                name: 'Test Person',
                titleRole: 'A Role',
                email: 'test+state+contact@example.com',
            },
        ],
        actuaryContacts: [],
        ...submissionPartial,
    }
}

const mockContractOnlyFormData = (
    submissionPartial?: Partial<LockedHealthPlanFormDataType>
): LockedHealthPlanFormDataType => {
    return {
        createdAt: new Date(),
        updatedAt: new Date(),
        status: 'SUBMITTED',
        stateNumber: 3,
        id: 'test-abc-125',
        stateCode: 'MN',
        programIDs: ['snbc'],
        submissionType: 'CONTRACT_ONLY',
        submissionDescription: 'A submitted submission',
        submittedAt: new Date(),
        documents: [
            {
                s3URL: 'bar',
                name: 'foo',
                documentCategories: ['RATES_RELATED' as const],
            },
        ],
        contractType: 'BASE',
        contractExecutionStatus: 'EXECUTED',
        contractDocuments: [
            {
                s3URL: 'bar',
                name: 'foo',
                documentCategories: ['CONTRACT' as const],
            },
        ],
        contractDateStart: new Date(),
        contractDateEnd: new Date(),
        managedCareEntities: ['ENROLLMENT_PROCESS'],
        federalAuthorities: ['VOLUNTARY', 'BENCHMARK'],
        rateDocuments: [],
        stateContacts: [
            {
                name: 'Test Person',
                titleRole: 'A Role',
                email: 'test+state+contact@example.com',
            },
        ],
        actuaryContacts: [],
        ...submissionPartial,
    }
}

const mockContractAmendmentFormData = (
    submissionPartial?: Partial<LockedHealthPlanFormDataType>
): LockedHealthPlanFormDataType => {
    return {
        createdAt: new Date(),
        updatedAt: new Date(),
        status: 'SUBMITTED',
        stateNumber: 3,
        id: 'test-abc-125',
        stateCode: 'MN',
        programIDs: ['snbc'],
        submissionType: 'CONTRACT_AND_RATES',
        submissionDescription: 'A submitted submission',
        submittedAt: new Date(),
        documents: [
            {
                s3URL: 'bar',
                name: 'foo',
                documentCategories: ['RATES_RELATED' as const],
            },
        ],
        contractType: 'AMENDMENT',
        contractExecutionStatus: 'UNEXECUTED',
        contractDocuments: [
            {
                s3URL: 'bar',
                name: 'foo',
                documentCategories: ['CONTRACT' as const],
            },
        ],
        contractDateStart: new Date(),
        contractDateEnd: new Date(),
        managedCareEntities: ['ENROLLMENT_PROCESS'],
        federalAuthorities: ['VOLUNTARY', 'BENCHMARK'],
        rateType: 'NEW',
        rateDocuments: [
            {
                s3URL: 'bar',
                name: 'foo',
                documentCategories: ['RATES' as const],
            },
        ],
        rateDateStart: new Date(),
        rateDateEnd: new Date(),
        rateDateCertified: new Date(),
        rateAmendmentInfo: null,
        stateContacts: [
            {
                name: 'Test Person',
                titleRole: 'A Role',
                email: 'test+state+contact@example.com',
            },
        ],
        actuaryContacts: [],
        ...submissionPartial,
    }
}

export {
    testEmailConfig,
    mockContractAmendmentFormData,
    mockContractOnlyFormData,
    mockContractAndRatesFormData,
    mockUnlockedContractAndRatesFormData,
    mockUnlockedContractOnlyFormData,
    mockUser,
    testEmailer,
}
