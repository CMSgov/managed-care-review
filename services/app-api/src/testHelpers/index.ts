export {
    assertAnError,
    assertAnErrorExtensions,
    assertAnErrorCode,
} from './gqlAssertions'

export { must } from './errorHelpers'

export {
    createInsertContractData,
    createContractData,
    createContractRevision,
    createDraftContractData,
} from './contractAndRates/contractHelpers'

export { getProgramsFromState, getStateRecord } from './stateHelpers'
