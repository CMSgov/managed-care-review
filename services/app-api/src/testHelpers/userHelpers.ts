import type { InsertUserArgsType } from '../postgres'
import { isStoreError, NewPostgresStore } from '../postgres'
import type {
    AdminUserType,
    CMSUserType,
    StateUserType,
    UserType,
} from '../domain-models'
import { sharedTestPrismaClient } from './storeHelpers'
import { v4 as uuidv4 } from 'uuid'

const testCMSUser = (userData?: Partial<CMSUserType>): CMSUserType => ({
    id: uuidv4(),
    role: 'CMS_USER',
    email: 'zuko@example.com',
    familyName: 'Zuko',
    givenName: 'Prince',
    divisionAssignment: 'DMCO' as const,
    stateAssignments: [],
    ...userData,
})

const testStateUser = (userData?: Partial<StateUserType>): StateUserType => ({
    id: uuidv4(),
    stateCode: 'FL',
    role: 'STATE_USER',
    email: 'james@example.com',
    familyName: 'Brown',
    givenName: 'James',
    ...userData,
})

const testAdminUser = (userData?: Partial<AdminUserType>): AdminUserType => ({
    id: uuidv4(),
    role: 'ADMIN_USER',
    email: 'iroh@example.com',
    familyName: 'Iroh',
    givenName: 'Uncle',
    ...userData,
})

const createDBUsersWithFullData = async (
    users: UserType[]
): Promise<UserType[]> => {
    const prismaClient = await sharedTestPrismaClient()
    const postgresStore = NewPostgresStore(prismaClient)

    const usersSeed: InsertUserArgsType[] = users.map((user: UserType) => {
        const userArgs: InsertUserArgsType & { id?: string } = {
            ...user,
            userID: user.id,
        }

        delete userArgs.id

        return userArgs as InsertUserArgsType
    })

    const result = await postgresStore.insertManyUsers(usersSeed)

    if (isStoreError(result)) {
        throw new Error(result.message)
    }

    return result
}

export { testAdminUser, testStateUser, testCMSUser, createDBUsersWithFullData }
