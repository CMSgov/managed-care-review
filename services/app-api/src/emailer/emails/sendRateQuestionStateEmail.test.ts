import { testCMSUser, testStateUser } from '../../testHelpers/userHelpers'
import {
    defaultFloridaProgram,
    defaultFloridaRateProgram,
} from '../../testHelpers/gqlHelpers'

import type { RateQuestionType, RateType } from '../../domain-models'
import { sendRateQuestionStateEmail } from './sendRateQuestionStateEmail'
import { testEmailConfig } from '../../testHelpers/emailerHelpers'

describe('sendRateQuestionStateEmail', () => {
    const dmcoCMSUser = testCMSUser({
        divisionAssignment: 'DMCO',
    })

    const currentQuestion = (): RateQuestionType => ({
        id: 'dmco-rate-question-1',
        rateID: 'test-rate',
        createdAt: new Date('2024-04-23'),
        addedBy: dmcoCMSUser,
        division: 'DMCO',
        documents: [
            {
                name: 'dmco-rate-question-1',
                s3URL: 's3://bucketName/key/dmco-rate-question-1-doc',
                downloadURL: 'https://fake-bucket.s3.amazonaws.com/test',
            },
        ],
        responses: [],
    })

    const testRate = (): RateType => ({
        id: 'test-rate',
        createdAt: new Date('2024-04-12'),
        updatedAt: new Date('2024-04-12'),
        status: 'SUBMITTED',
        stateCode: 'FL',
        parentContractID: 'parent-contract',
        stateNumber: 2,
        revisions: [],
        packageSubmissions: [
            {
                submitInfo: {
                    updatedAt: new Date('2024-04-12'),
                    updatedReason: 'initial submission',
                    updatedBy: testStateUser({
                        email: 'parent-contract-submitter@state.com',
                    }),
                },
                submittedRevisions: [],
                rateRevision: {
                    id: 'test-rate-revision',
                    rateID: 'test-rate',
                    submitInfo: {
                        updatedAt: new Date('2024-04-12'),
                        updatedReason: 'initial submission',
                        updatedBy: testStateUser({
                            email: 'parent-contract-submitter@state.com',
                        }),
                    },
                    createdAt: new Date('2024-04-12'),
                    updatedAt: new Date('2024-04-12'),
                    formData: {
                        rateType: 'NEW',
                        rateCapitationType: 'RATE_CELL',
                        rateDateStart: new Date('2024-01-01'),
                        rateDateEnd: new Date('2025-01-01'),
                        rateDateCertified: new Date('2024-01-02'),
                        rateCertificationName: 'test-rate-certification-name',
                        rateProgramIDs: [defaultFloridaRateProgram().id],
                        rateDocuments: [
                            {
                                s3URL: 's3://bucketname/key/test1',
                                name: 'ratedoc1.doc',
                                sha256: 'foobar',
                            },
                        ],
                        supportingDocuments: [],
                        certifyingActuaryContacts: [
                            {
                                name: 'Foo Person',
                                titleRole: 'Bar Job',
                                email: 'certifyingActuary@example.com',
                                actuarialFirm: 'GUIDEHOUSE',
                            },
                        ],
                        addtlActuaryContacts: [
                            {
                                name: 'Bar Person',
                                titleRole: 'Baz Job',
                                email: 'addtlActuaryContacts@example.com',
                                actuarialFirm: 'OTHER',
                                actuarialFirmOther: 'Some Firm',
                            },
                        ],
                        actuaryCommunicationPreference: 'OACT_TO_ACTUARY',
                    },
                },
                contractRevisions: [
                    {
                        id: 'parent-contract-revision',
                        contract: {
                            id: 'parent-contract',
                            stateCode: 'FL',
                            stateNumber: 1,
                        },
                        submitInfo: {
                            updatedAt: new Date('2024-04-12'),
                            updatedReason: 'initial submission',
                            updatedBy: testStateUser({
                                email: 'parent-contract-submitter@state.com',
                            }),
                        },
                        createdAt: new Date('2024-04-12'),
                        updatedAt: new Date('2024-04-12'),
                        formData: {
                            programIDs: [defaultFloridaProgram().id],
                            submissionDescription: 'submission description',
                            submissionType: 'CONTRACT_AND_RATES',
                            contractType: 'BASE',
                            contractExecutionStatus: 'EXECUTED',
                            contractDateStart: new Date('2024-01-01'),
                            contractDateEnd: new Date('2025-01-01'),
                            contractDocuments: [],
                            managedCareEntities: ['MCO'],
                            federalAuthorities: ['STATE_PLAN'],
                            populationCovered: 'MEDICAID',
                            supportingDocuments: [],
                            stateContacts: [
                                {
                                    name: 'parent-contract-state-contact',
                                    titleRole: 'state contact',
                                    email: 'parent-contract-state-contact-1@state.com',
                                },
                                {
                                    name: 'duplicate-state-contact',
                                    titleRole: 'state contact',
                                    email: 'duplicateContact@state-contact.com',
                                },
                                {
                                    name: 'duplicate-state-contact',
                                    titleRole: 'state contact',
                                    email: 'duplicateContact@state-contact.com',
                                },
                            ],
                        },
                    },
                    {
                        id: 'second-contract-revision',
                        contract: {
                            id: 'second-contract',
                            stateCode: 'FL',
                            stateNumber: 1,
                        },
                        submitInfo: {
                            updatedAt: new Date('2024-04-12'),
                            updatedReason: 'initial submission',
                            updatedBy: testStateUser({
                                email: 'second-contract-submitter@state.com',
                            }),
                        },
                        createdAt: new Date('2024-04-12'),
                        updatedAt: new Date('2024-04-12'),
                        formData: {
                            programIDs: [defaultFloridaProgram().id],
                            submissionDescription: 'submission description',
                            submissionType: 'CONTRACT_AND_RATES',
                            contractType: 'BASE',
                            contractExecutionStatus: 'EXECUTED',
                            contractDateStart: new Date('2024-01-01'),
                            contractDateEnd: new Date('2025-01-01'),
                            contractDocuments: [],
                            managedCareEntities: ['MCO'],
                            federalAuthorities: ['STATE_PLAN'],
                            populationCovered: 'MEDICAID',
                            supportingDocuments: [],
                            stateContacts: [
                                {
                                    name: 'second-contract-state-contact',
                                    titleRole: 'state contact',
                                    email: 'second-contract-state-contact-1@state.com',
                                },
                                {
                                    name: 'duplicate-state-contact',
                                    titleRole: 'state contact',
                                    email: 'duplicateContact@state-contact.com',
                                },
                                {
                                    name: 'duplicate-state-contact',
                                    titleRole: 'state contact',
                                    email: 'duplicateContact@state-contact.com',
                                },
                            ],
                        },
                    },
                ],
            },
        ],
        questions: {
            DMCOQuestions: {
                totalCount: 1,
                edges: [
                    {
                        node: currentQuestion(),
                    },
                ],
            },
            DMCPQuestions: {
                totalCount: 0,
                edges: [],
            },
            OACTQuestions: {
                totalCount: 0,
                edges: [],
            },
        },
    })

    it('to addresses list includes submitter emails', async () => {
        const template = await sendRateQuestionStateEmail(
            testRate(),
            testEmailConfig(),
            currentQuestion()
        )

        if (template instanceof Error) {
            throw template
        }

        expect(template).toEqual(
            expect.objectContaining({
                toAddresses: expect.arrayContaining([
                    'parent-contract-submitter@state.com',
                    'second-contract-submitter@state.com',
                ]),
            })
        )
    })
    it('to addresses list includes all state contacts on all contracts submitted with rate', async () => {
        const template = await sendRateQuestionStateEmail(
            testRate(),
            testEmailConfig(),
            currentQuestion()
        )

        if (template instanceof Error) {
            throw template
        }

        expect(template).toEqual(
            expect.objectContaining({
                toAddresses: expect.arrayContaining([
                    'parent-contract-state-contact-1@state.com',
                    'duplicateContact@state-contact.com',
                    'second-contract-state-contact-1@state.com',
                    'certifyingActuary@example.com',
                    'addtlActuaryContacts@example.com',
                ]),
            })
        )
    })
    it('to addresses does not include actuaries when rate communication preference is OACT_TO_STATE', async () => {
        const rate = testRate()
        rate.packageSubmissions[0].rateRevision.formData.actuaryCommunicationPreference =
            'OACT_TO_STATE'
        const template = await sendRateQuestionStateEmail(
            rate,
            testEmailConfig(),
            currentQuestion()
        )

        if (template instanceof Error) {
            throw template
        }

        expect(template).toEqual(
            expect.objectContaining({
                toAddresses: expect.arrayContaining([
                    'parent-contract-state-contact-1@state.com',
                    'duplicateContact@state-contact.com',
                    'second-contract-state-contact-1@state.com',
                ]),
            })
        )
    })
    it('to addresses list does not include duplicate emails', async () => {
        const template = await sendRateQuestionStateEmail(
            testRate(),
            testEmailConfig(),
            currentQuestion()
        )

        if (template instanceof Error) {
            throw template
        }

        // filter out known duplicate emails
        const filterKnownDuplicateEmails = template.toAddresses.filter(
            (email) => email === 'duplicateContact@state-contact.com'
        )

        // expect only 1 of the duplicate emails in the toAddress
        expect(filterKnownDuplicateEmails).toHaveLength(1)
    })
    it('subject line is correct and clearly states submission is complete', async () => {
        const template = await sendRateQuestionStateEmail(
            testRate(),
            testEmailConfig(),
            currentQuestion()
        )

        if (template instanceof Error) {
            throw template
        }

        expect(template).toEqual(
            expect.objectContaining({
                subject: expect.stringContaining(
                    `New questions about test-rate-certification-name`
                ),
                bodyText: expect.stringContaining(
                    'test-rate-certification-name'
                ),
            })
        )
    })
    it('includes link to rate Q&A page', async () => {
        const template = await sendRateQuestionStateEmail(
            testRate(),
            testEmailConfig(),
            currentQuestion()
        )

        if (template instanceof Error) {
            throw template
        }

        expect(template).toEqual(
            expect.objectContaining({
                bodyText: expect.stringContaining(
                    'Open the submission in MC-Review to answer question'
                ),
                bodyHTML: expect.stringContaining(
                    `http://localhost/submissions/parent-contract/rates/test-rate/question-and-answer`
                ),
            })
        )
    })
    it('includes expected data on the CMS analyst who sent the question', async () => {
        const template = await sendRateQuestionStateEmail(
            testRate(),
            testEmailConfig(),
            currentQuestion()
        )

        if (template instanceof Error) {
            throw template
        }

        expect(template).toEqual(
            expect.objectContaining({
                bodyText: expect.stringContaining(
                    'Sent by: Prince Zuko (DMCO)  zuko@example.com (zuko@example.com)'
                ),
            })
        )
        expect(template).toEqual(
            expect.objectContaining({
                bodyText: expect.stringContaining('Date: 04/22/2024'),
            })
        )
    })
    it('renders overall email for a new rate question as expected', async () => {
        const template = await sendRateQuestionStateEmail(
            testRate(),
            testEmailConfig(),
            currentQuestion()
        )

        if (template instanceof Error) {
            throw template
        }

        expect(template).toEqual(
            expect.objectContaining({
                bodyText: expect.stringContaining(
                    'Sent by: Prince Zuko (DMCO)  zuko@example.com (zuko@example.com)'
                ),
            })
        )
        expect(template.bodyHTML).toMatchSnapshot()
    })
})
