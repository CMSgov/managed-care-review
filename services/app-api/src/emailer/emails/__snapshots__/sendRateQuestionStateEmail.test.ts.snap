// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sendRateQuestionStateEmail renders overall email for a new rate question as expected 1`] = `
"CMS asked questions about test-rate-certification-name<br />
<b>Sent by:</b> Prince Zuko (DMCO) <a href="zuko@example.com">zuko@example.com</a>
<br />
<b>Date:</b> 04/22/2024<br />
<br />
You must respond to the questions before CMS can continue their review. Responses should be submitted in DOC or DOCX format.<br />
<br />
<a href="http://localhost/submissions/parent-contract/rates/test-rate/question-and-answers">Open the submission in MC-Review to answer questions</a>

"
`;
